File: octave/ch1/Ex122.m
function Ex122
  #search_for_primes(1000, 1020);
  search_for_primes(10000, 10040);
  #search_for_primes(100000, 100080);
  printf("done\n");
endfunction


function VAL = find_divisor(n, test)
  divides = @(a, b) rem(b, a) == 0;
  
  if (test * test > n)
    VAL = n;
  elseif (divides(test, n))
    VAL = test;
  else
    VAL = find_divisor(n, test + 1);
  endif
endfunction

function VAL = smallest_divisor(n) 
  VAL = find_divisor(n, 2);
endfunction

function VAL = search_for_primes(start, finish)
  n = start;
  if rem(start, 2) == 0
    n = n + 1;
  endif
  
  VAL = search_helper(n, finish);
endfunction

function VAL = search_helper(n, finish)
  range = n:2:finish;
  results = arrayfun(@prime, range);
  
  if n <= finish
    printf("%d %d\n", [range; results]);
  else
    printf("finished\n");
  endif
  
  VAL = finish;
endfunction

function VAL = prime(n)
  VAL = n == smallest_divisor(n);
endfunction

File: rlang/ch1/ex122.R
# setwd("c:/Users/Yohan/IdeaProjects/sicp/rlang/ch1")
# getwd()
# load with source("filename")


divides <- function(a, b) {
  0 == b %% a
}

find_divisor <- function(n, test) {
  if (test * test > n) {
    n
  } else if (divides(test, n)) {
    test
  } else {
    find_divisor(n, test + 1)
  }
}

smallest_divisor <- function(n) {
  find_divisor(n, 2)
}

is_prime <- function(n) {
  n == smallest_divisor(n)
}

report_prime <- function(n) {
  result = ifelse(is_prime(n), " *** ", "")
  print(noquote(paste(toString(n), result)))
}

search_helper <- function(n, end) {
  if (n <= end) {
    report_prime(n)
    search_helper(n + 2, end)
  } else {
    print(noquote("finished"))
  }
}

search_for_primes <- function(start, end) {
  search_helper(ifelse(divides(2, start), start + 1, start), end)
}


# Run in console
# > library(testthat)
# > test_file('this_file.R')

# test_that("test1", {
#   expect_equal(smallest_divisor(199), 199)
#   expect_equal(smallest_divisor(1999), 1999)
#   expect_equal(smallest_divisor(19999), 7)
# })

# Try these tests
search_for_primes(1000, 1020)
search_for_primes(10000, 10040)
search_for_primes(100000, 100080)


